pkg/rate/api_limiter.go

```diff
@@ -0,0 +1,790 @@
+	// AutoAdjust is enabled, the value will adjust automatically.
+	RateBurst int
+
+	// MinWaitDuration is the minimum time an API request always has to
+	// wait before the Wait() function returns an error.
+	MinWaitDuration time.Duration
+
+	// MaxWaitDuration is the maximum time an API request is allowed to
+	// wait before the Wait() function returns an error.
+	MaxWaitDuration time.Duration
+
+	// Log enables info logging of processed API requests. This should only
+	// be used for low frequency API calls.
+	Log bool
+
+	// DelayedAdjustmentFactor is percentage of the AdjustmentFactor to be
+	// applied to RateBurst and MaxWaitDuration defined as a value between
+	// 0.0..1.0. This is used to steer a slower reaction of the RateBurst
+	// and ParallelRequests compared to RateLimit.
+	DelayedAdjustmentFactor float64
+
+	// SkipInitial is the number of initial API calls for which to not
+	// apply any rate limiting. This is useful to define a learning phase
+	// in the beginning to allow for auto adjustment before imposing wait
+	// durations and rate limiting on API calls.
+	SkipInitial int
+
+	// MaxAdjustmentFactor is the maximum adjustment factor when AutoAdjust
+	// is enabled. Base values will not adjust more than by this factor.
+	MaxAdjustmentFactor float64
+}
+
+// MergeUserConfig merges the provided user configuration into the existing
+// parameters and returns a new copy.
+func (p APILimiterParameters) MergeUserConfig(config string) (APILimiterParameters, error) {
+	c := p
+
+	if err := parseUserConfig(config, &c); err != nil {
+		return APILimiterParameters{}, err
+	}
+
+	return c, nil
+}
+
+// NewAPILimiter returns a new APILimiter based on the parameters and metrics implementation
+func NewAPILimiter(name string, p APILimiterParameters, metrics MetricsObserver) *APILimiter {
+	if p.MeanOver == 0 {
+		p.MeanOver = defaultMeanOver
+	}
+
+	if p.MinParallelRequests == 0 {
+		p.MinParallelRequests = 1
+	}
+
+	if p.RateBurst == 0 {
+		p.RateBurst = 1
+	}
+
+	if p.DelayedAdjustmentFactor == 0.0 {
+		p.DelayedAdjustmentFactor = defaultDelayedAdjustmentFactor
+	}
+
+	if p.MaxAdjustmentFactor == 0.0 {
+		p.MaxAdjustmentFactor = defaultMaxAdjustmentFactor
+	}
+
+	l := &APILimiter{
+		name:                  name,
+		params:                p,
+		parallelRequests:      p.ParallelRequests,
+		parallelWaitSemaphore: semaphore.NewWeighted(waitSemaphoreResolution),
+		metrics:               metrics,
+	}
+
+	if p.RateLimit != 0 {
+		l.limiter = rate.NewLimiter(p.RateLimit, p.RateBurst)
+	}
+
+	return l
+}
+
+// NewAPILimiterFromConfig returns a new APILimiter based on user configuration
+func NewAPILimiterFromConfig(name, config string, metrics MetricsObserver) (*APILimiter, error) {
+	p := APILimiterParameters{}
+
+	if err := parseUserConfig(config, &p); err != nil {
+		return nil, err
+	}
+
+	return NewAPILimiter(name, p, metrics), nil
+}
+
+func parseUserConfigKeyValue(key, value string, p *APILimiterParameters) error {
```

Comment:
Seems like this could/should use receivers instead? Was there a reason to avoid this?

```suggestion
func (p *APILimiterParameters) parseUserConfigKeyValue(key, value string) error {
```

(not the only case, and probably not such a big deal, just a little unusual for Go code)

Reply:
It's written like this because no other function of APILimiterParameters modifies the struct so it's clearly immutable. I worry that by adding a receiver version, somebody will assume concurrent access is OK by accident. By having to pass in a pointer, it's very obvious that concurrent access is not automatically taken care of.

Reply:
I think documenting this function to state that it doesn't provide concurrent-safe access is preferred. Then it would be fine to have it be a receiver. The way it is being passed currently, it's essentially what the compiler already does under-the-hood. IMO, documenting the function is a stronger way to signal the intent, over relying on the reader's ability to understand why we are passing via args explicitly.

Reply:
Sure, I won't insist on this. Changed.